src/sampling_timer.rs

```diff
@@ -0,0 +1,147 @@
+///! The sampling timer is used for managing ADC sampling and external reference timestamping.
+use super::hal;
+
+pub use hal::stm32::tim2::ccmr2_input::CC4S_A;
+
+/// The timer used for managing ADC sampling.
+pub struct SamplingTimer {
+    timer: hal::timer::Timer<hal::stm32::TIM2>,
+    channels: Option<TimerChannels>,
+}
+
+impl SamplingTimer {
+    /// Construct the sampling timer.
+    pub fn new(mut timer: hal::timer::Timer<hal::stm32::TIM2>) -> Self {
+        timer.pause();
+
+        Self {
+            timer,
+            channels: Some(TimerChannels::new()),
+        }
+    }
+
+    /// Get the timer capture/compare channels.
+    pub fn channels(&mut self) -> TimerChannels {
+        self.channels.take().unwrap()
+    }
```

Comment:
This holds because the HAL is not currently making use of any of the timer channels. Ideally, this type of support would be added to the HAL in the future, although you make a good point.

If the HAL were updated in the future to start using the underlying timer channels, invariants may break. I don't see a way around this beyond documenting it in the code.

Reply:
Added docs throughout about safety.